### ISSUE TITLE
Introduce a way to relax/disable compiler version checking against settings.yml

### ISSUE THREAD

**Issue by unjello** [conan-io/conan#2650]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

---

FWIW my environment is:
```bash
❯ conan --version
Conan version 1.1.1
❯ uname -a
Darwin local 17.4.0 Darwin Kernel Version 17.4.0: Sun Dec 17 09:19:54 PST 2017; root:xnu-4570.41.2~1/RELEASE_X86_64 x86_64
```
The CI worker environment is:
```bash
Worker information
hostname: f5c14a39-5e7c-4f30-b11c-6b2983c46c5c@1.production-1-worker-org-c-2-gce
version: v3.6.0 https://github.com/travis-ci/worker/tree/170b2a0bb43234479fd1911ba9e4dbcc36dadfad
instance: travis-job-9b4e7152-aa1f-4511-a05d-3d02109082ba travis-ci-garnet-trusty-1512502259-986baf0 (via amqp)
```

although the issue is OS agnostic.

1. Set up environment
```bash
❯ export COMPILER_NAME=clang
❯ export COMPILER_VERSION=6.0
```

2. Run `conan install`:
```bash
$ conan install . --build missing -s compiler=$COMPILER_NAME -s compiler.version=$COMPILER_VERSION
Auto detecting your dev setup to initialize the default profile (/home/travis/.conan/profiles/default)
CC and CXX: clang-6.0, clang++-6.0 
Found clang 6.0
Default settings
	os=Linux
	os_build=Linux
	arch=x86_64
	arch_build=x86_64
	compiler=clang
	compiler.version=6.0
	compiler.libcxx=libstdc++
	build_type=Release
*** You can change them in /home/travis/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***
ERROR: Invalid setting '6.0' is not a valid 'settings.compiler.version' value.
Possible values are ['3.3', '3.4', '3.5', '3.6', '3.7', '3.8', '3.9', '4.0', '5.0']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"
The command "conan install . --build missing -s compiler=$COMPILER_NAME -s compiler.version=$COMPILER_VERSION" failed and exited with 1 during .
Your build has been stopped.
```

When working with versions of compilers that are not pre-configured in `settings.yml`, ie. `clang 6.0` or `svn` builds of `gcc` and `clang` that may have custom version strings, especially in CI environment, it would be great if there was a flag that I could pass to Conan to bypass validation against `settings.yml`. The workaround is to checkin my `settings.yml` and in CI builds copy it over to `~/.conan/settings.yml` which is a bit cumbersome.

I would propose to introduce a flag and environment variable to relax/disable failure if compiler version is not present in `settings.yml`. I think checking for name could be left in, as it's more fundamental to how Conan works.

**Comment by lasote** [comment]
I don't think it should be done. For me, the gain does not justify the mess you can generate if you don't validate the settings against something known.

**Comment by danimtb** [comment]
I think this kind of "brakes" the conan validation of settings and the input of the consumers. The best way would be to add it to your own *settings.yml* via [conan config isntall](http://docs.conan.io/en/latest/reference/commands/consumer/config.html#conan-config-install) or something like that.

BTW, clang 6.0 support will be added for next release. Check https://github.com/conan-io/conan/pull/2591

**Comment by unjello** [comment]
@lasote For you, maybe. But since I can put anything into handcrafted `settings.yml` anyway, what are we validating agains? I mean, I will put that `6.0` there one way or another. Currently it is just cumbersome for me to achieve what I want.

@danimtb While I think there are legitimate scenarios for full control and strict validation, I think there are also legitimate cases where I would like Conan to *validate if you can fetch pre-built, if not try to build, and fail only if build fails*. Sure, Clang 6.0 will be in next release, but when will you add 7.0? Or how to handle gracefully svn builds for testing? It is a common practice in Golang to test against `tip` release. C++ could use the same thing.

Either way, how about a middle ground? If `.conan-settings.yml` is present in current working directory, it overrides `~/.conan/settings.yml`? This way, we have a compromise, such you keep Conan validation (which again, I could argue isn't that useful if I can put anything into settings myself) but saves me from having to do extra steps in CI workers, and keeps my repo cleaner, as dot-files are common pattern for storing tools-related configuration options for repo.

**Comment by lasote** [comment]
> But since I can put anything into handcrafted settings.yml anyway, what are we validating against? 

Against the values introduced but you, of course. But eventually validating. Not "6" nor "6.0." nor "six" would be valid, only "6.0"

**Comment by unjello** [comment]
> Against the values introduced by you, of course. But eventually validating. Not "6" nor "6.0." nor "six" would be valid, only "6.0"

so if there's another stage of validation, why test if the version is present in `settings.yml`? Why not do it such that I pass `-s compiler.version=6.0` and you don't check if that is present in `settings.yml`, just validate that it is not `six`?

> You can change your conan folder to another place with an env var `CONAN_USER_HOME`

It's a bit better, but still not ideal. It forces me to have a `./conan/settings.yml` tree in my repo, also it'll eventually store all `conan` related data in there during a build, which makes caching between CI worker runs a bit harder... it is a common practice to allow for granularity of configuration, from environment variables to local dot-config files, to per-user files and eventually system config files, to some extent described in [12 factor app](https://12factor.net) manifesto.

**Comment by niosHD** [comment]
I am in favor for having an (opt-in option) to disable validation against the `settings.yml`. I, for example, do CI testing of bleeding edge compilers (i.e., builds from trunk/master). The compiler version can change with every build and the only way we currently have to cope with it is to generate a new `settings.yml` which contains the version the compiler identifies as.

In this usage scenario, the validation does not provide any benefit at all but only is holding me back. As @unjello highlighted, I can put anything into the `settings.yml` and that is exactly what have to do right now to make it work.

In summary, I absolutely see the value of the validation in general, but for certain use cases like @unjello's and mine, it is in the way. What is the harm of letting the user opt-out of it?

**Comment by lasote** [comment]
In general, guys, there is always a use case, totally legit sometimes, but keeping a reasonable decent codebase includes not introducing all the use cases, especially if it affects the core model.
I really think that removing the validation is not the way to go. So we are not doing this at the moment. I keep this issue opened, of course, any more feedback is welcome.

**Comment by memsharded** [comment]
I would like to add something to the conversation. Even if it is not my preferred strategy, but due to users requests, there is more and more "intelligence" added specially to the build helpers.

Please check the ongoing work here: https://github.com/conan-io/conan/pull/2646/files

In this case, the clang 6.0 version affects in 2 different places:
- It is necessary to change the default ``cppstd`` to c++14, which is the one adopted by clang 6, which affects the binary model
- It is necessary to add adequate ``-std=c++17`` compiler flags, that affects build helpers.

So in general, using free-form settings can produce undefined results, specially if using the predefined build helpers, but also regarding the binaries model (having more binaries than necessary, for example)

It is common that users that are customizing ``settings.yml`` they would also have their own build helpers customized for their use cases, so if they are adding custom settings, they can modify their build helpers accordingly.

A possible very simple and pragmatic approach would be to define ``version=ANY`` in your own custom ``settings.yml``. In that way, you don't need to adapt at all in each build, you will be always doing a ``conan config install`` of your settings, and it will be useful forever, for any versions, etc. (of course the only caution is to know about the above mentioned limitations of build helpers and binary model)

**Comment by niosHD** [comment]
Thank you @memsharded for the hint with `version: ANY`, I completely missed this possibility. I still have to deploy a custom `settings.yml` due to custom architectures. However, they do by far not change that often.

**Comment by unjello** [comment]
@memsharded huge props for sharing this trick. This does work for me. Not the prettiest solution for my needs, but good enough, given constraints that drive conan design decisions.

**Comment by memsharded** [comment]
You are welcome, glad that it is good enough. Thanks for all your feedback!
